doc/neps/nep-0049.rst

```diff
@@ -0,0 +1,338 @@
+===================================
+NEP 49 — Data allocation strategies
+===================================
+
+:Author: Matti Picus
+:Status: Draft
+:Type: Standards Track
+:Created: 2021-04-18
+
+
+Abstract
+--------
+
+The ``numpy.ndarray`` requires additional memory allocations
+to hold ``numpy.ndarray.strides``, ``numpy.ndarray.shape`` and
+``numpy.ndarray.data`` attributes. These attributes are specially allocated
+after creating the python object in ``__new__`` method. The ``strides`` and
+``shape`` are stored in a piece of memory allocated internally.
+
+These allocations are small relative to the ``data``, the homogeneous chunk of
+memory used to store the actual array values (which could be pointers in the
+case of ``object`` arrays). This NEP proposes a mechanism to override the
+default memory management strategy in NumPy with user-provided alternatives.
+
+Motivation and Scope
+--------------------
+
+Users may wish to override the internal data memory routines with ones of their
+own. Two such use-cases are to ensure data alignment and to pin certain
+allocations to certain NUMA cores. The NEP is based on `PR 17582`_ and
+largely taken from the documentation added there.
+
+User who wish to change the NumPy data memory management routines will use
+:c:func:`PyDataMem_SetHandler`, which uses a :c:type:`PyDataMem_Handler`
+structure to hold pointers to functions used to manage the data memory. The
+calls are wrapped by internal routines to call :c:func:`PyTraceMalloc_Track`,
+:c:func:`PyTraceMalloc_Untrack`, and will use the
+:c:func:`PyDataMem_EventHookFunc` mechanism  already present in NumPy for
+auditing purposes. 
+
+Since a call to ``PyDataMem_SetHandler`` will change the default functions, but
+that function may be called during the lifetime of an ``ndarray`` object, each
+``ndarray`` will carry with it the ``PyDataMem_Handler`` struct used at the
+time of its instantiation, and these will be used to reallocate or free the
+data memory of the instance.
+
+Usage and Impact
+----------------
+
+The new functions can only be accessed via the NumPy C-API. An example is
+included later in the NEP. The added ``struct`` will increase the size of the
+``ndarray`` object. It is one of the major drawbacks of this approach. We can
+be reasonably sure that the change in size will have a minimal impact on
+end-user code because NumPy version 1.20 already changed the object size.
```

Comment:
Seems fine here, in line with my questions below, it might be nice to have one paragraph that explains how users will benefit (that is probably the users of the libraries that modify the allocation strategy).

I assume this are things like: Alignment guarantees or NUMA core pinning can speed up their code. Allocating in memory shared by GPU and CPU speeds up certain operations?

Reply:
> Alignment guarantees or NUMA core pinning can speed up their code

Yes. There are libraries like `pnumpy` that would like to keep data on a specific core or hardware where unaligned memory is expensive (like ARMv7).

Reply:
I think using pinned memory would hit some snags due to `memcpy`. A future enhancement might be to enable user-supplied `memcpy` routines and using them inside numPy when creating a copy of an ndarray.

Reply:
Just FYI. CuPy recently added a few APIs to allocate pinned memory that backs a NumPy array: 
> For using pinned memory more conveniently, we also provide a few high-level APIs in the cupyx namespace, including cupyx.empty_pinned(), cupyx.empty_like_pinned(), cupyx.zeros_pinned(), and cupyx.zeros_like_pinned(). They return NumPy arrays backed by pinned memory. If CuPy’s pinned memory pool is in use, the pinned memory is allocated from the pool.

https://docs.cupy.dev/en/stable/user_guide/memory.html#memory-management. I think PyCUDA also have similar APIs. 

Indeed, memcpy could be an issue for D2H transfers, but not H2D (at least in CuPy, `arr_cp[:] = arr_np_on_pinned` would handle this properly). It's something I am working to circumvent (by calling `arr_np_on_pinned = cp.asnumpy(arr_cp, out=arr_np_on_pinned)`).